File: python/organ_donations/dslc_documentation/functions/impute_feature.py
def impute_feature(data, feature, group, impute_method="average"):
    impute_method = impute_method.lower()
    
    if impute_method == "previous":
        data = (data.assign(feature_imputed=data[feature]) # create a new variable called feature_imputed that is equal to feature
                    .groupby(group) # group by 
                    .fillna(method='ffill') # impute using forward fill
                    .fillna(0)) # impute remaining missing values with 0
        return data["feature_imputed"]
    elif impute_method == "average":
        # create two temporary variables each equal to feature
        data = data.assign(imputed_feature_tmp_prev=data[feature], imputed_feature_tmp_next=data[feature])
        # fill the first variable using forward fill
        data["imputed_feature_tmp_prev"] = data.groupby(group)["imputed_feature_tmp_prev"].fillna(method='ffill')
        # fill the second variable using backward fill
        data["imputed_feature_tmp_next"] = data.groupby(group)["imputed_feature_tmp_next"].fillna(method='bfill')
        # then define feature_imputed column to be the mean of the forward and backward filled values
        data['feature_imputed'] = data[['imputed_feature_tmp_next', 'imputed_feature_tmp_prev']].mean(axis=1, skipna=True)
        # impute any remaining missing values with 0
        data['feature_imputed'].fillna(0, inplace=True)
        # remove the two temporary variables
        data = data.drop(columns=['imputed_feature_tmp_prev', 'imputed_feature_tmp_next'])
        return data["feature_imputed"]
    else:
        raise ValueError

File: R/organ_donations/dslc_documentation/functions/imputeFeature.R

# function for conducting the imputation of a specified feature
imputeFeature <- function(.data,
                          .feature,
                          .group,
                          .impute_method = c("average", "previous")) {
  # identify which imputation method is being used
  .impute_method <- match.arg(.impute_method)
  
  if (.impute_method == "previous") {
    .data <- .data |>
      # duplicate the relevant column and call it "feature_imputed"
      # this {{ }} notation is "tidy eval" - it allows us to pass unquoted 
      # variable names
      mutate(feature_imputed = {{ .feature }}) |>
      # for each .group (country)
      group_by({{ .group }}) |>
      # use the fill function from the tidyr package (part of the tidyverse)
      # to impute the relevant feature
      fill(feature_imputed, .direction = "down") |>
      ungroup() |>
      # for entries whose values are still missing (e.g., countries with no 
      # data reported), fill the missing value with 0
      mutate(feature_imputed = if_else(is.na(feature_imputed), 0, feature_imputed))
  } else if (.impute_method == "average") {
    .data <- .data |>
      # duplicate the relevant column two times and call these duplicates
      # "feature_imputed_tmp_prev" and "feature_imputed_tmp_next"
      # we will fill the missing values in these columns with the previous and 
      # next non-missing values respectively
      # this {{ }} notation is "tidy eval" - it allows us to pass unquoted 
      # variable names
      mutate(imputed_feature_tmp_prev = {{ .feature }},
             imputed_feature_tmp_next = {{ .feature }}) |>
      # for each .group (country)
      group_by({{ .group }}) |>
      # use the fill function to impute the missing values with previous 
      # non-missing value 
      fill(imputed_feature_tmp_prev, .direction = "down") |>
      # use the fill function to impute the missing values with next 
      # non-missing value
      fill(imputed_feature_tmp_next, .direction = "up") |>
      ungroup() |>
      # compute the imputation as the average of the previous and next imputed 
      # value for each row.
      rowwise() |>
      mutate(feature_imputed = mean(c(imputed_feature_tmp_next, 
                                    imputed_feature_tmp_prev), na.rm = T)) |>
      # for entries whose values are still missing (e.g., countries with no 
      # data reported), fill the missing value with 0
      mutate(feature_imputed = if_else(is.nan(feature_imputed), 0, feature_imputed)) |>
      # remove the two prev, next columns
      select(-imputed_feature_tmp_prev, -imputed_feature_tmp_next)
  }
  # return just the imputed feature
  return(pull(.data, feature_imputed))
}
